geograypher/utils/image.py

```diff
@@ -19,3 +24,142 @@ def get_GPS_exif(filename):
+        yaw_deg (float, optional): yaw camera orientation. Defaults to 0.
+        pitch_deg (float, optional): pitch camera orientation. Defaults to 0.
+        roll_deg (float, optional): roll camera orientation. Defaults to 0.
+        warp_order (int, optional): Interpolation order to use when sampling pixels. Defaults to 1.
+        oversample_factor (int, optional):
+            Sample this number times more pixels in each dimension. This helps avoid aliasing but
+            increases runtime. Defaults to 1.
+        return_mask (bool, optional):
+           Return a mask showing what pixels from the original image were
+             sampled. Defaults to False.
+
+    Returns:
+        np.ndarray: The sampled perspective image
+        Optional[np.ndarray]]: The mask of sampled locations, if requested
+    """
+    # Convert an equirectangular (360) image to a perspective view.
+    H, W = equi_img.shape[:2]
+
+    # Output image dimensions
+    out_w, out_h = output_size
+
+    # Sample a larger image and then downsample to reduce aliasing
+    out_w = int(out_w * oversample_factor)
+    out_h = int(out_h * oversample_factor)
+
+    # FOV and angles in radians
+    fov = np.deg2rad(fov_deg)
+    yaw = np.deg2rad(yaw_deg)
+    pitch = np.deg2rad(pitch_deg)
+    roll = np.deg2rad(roll_deg)
+
+    # Compute the aspect ratio of the requested image dimensions
+    aspect_ratio = out_h / out_w
+
+    # Create homogenous image coordinates for the output image
+    x = np.linspace(-np.tan(fov / 2), np.tan(fov / 2), out_w)
+    y = np.linspace(
+        -np.tan(fov / 2) * aspect_ratio, np.tan(fov / 2) * aspect_ratio, out_h
+    )
+
+    xv, yv = np.meshgrid(
+        x, -y
+    )  # invert y for image coordinates to account for array indexing notation
+
+    # The z direction
+    zv = np.ones_like(xv)
+
+    pixel_directions = np.stack((xv, yv, zv), axis=-1)
+    # normalize to unit
+    pixel_directions /= np.linalg.norm(pixel_directions, axis=-1, keepdims=True)
+
+    # Seems to correspond to the roll-pitch-yaw convention
+    rotation_matrix = Rotation.from_euler("zyx", [roll, yaw, pitch]).as_matrix()
+
+    # Rotate the pixel directions by the rotation matrix
+    # The strange convention here is to deal with the fact that pixel_directions is (w, h, 3)
+    # so this allows the dimensions to align for the matrix multiplication.
+    pixel_directions = pixel_directions @ rotation_matrix.T
+
+    # Convert 3D directions to spherical coordinates
+    # horizontal angle
+    horizontal = np.arctan2(pixel_directions[..., 0], pixel_directions[..., 2])
+    # vertical angle
+    altitude = np.arcsin(pixel_directions[..., 1])
+
+    # Map to equirectangular image coordinates
+    i = (0.5 - altitude / np.pi) * H
+    j = (horizontal / (2 * np.pi) + 0.5) * W
+
+    # Replicate the pixels from the opposite side of the image to handle wrap-around
+    equi_img = np.concatenate(
+        [equi_img[:, -1:, :], equi_img, equi_img[:, 0:1, :]], axis=1
```

Comment:
This is a little confusing, what's going on? Is j expected to be negative such that we're worried about the wraparound? Or is it expected to be >= W?

Reply:
I tried to document this better. If a value is greater than W - 1, which could happen due to the scaling, we should be partially sampling from the leftmost row if using interpolation. This allows this to happen by padding the image which is being sampled from.